haddock-api/src/Haddock/Backends/Annot.hs

```diff
@@ -0,0 +1,185 @@
+    where target = odir </> file
+
+render :: Handle -> Interface -> IO ()
+render annH iface
+  = do putStrLn $ "Render Annot: src = " ++ src ++ " module = " ++ modl
+       hPutStr annH annots
+    where src     = ifaceOrigFilename iface
+          modl    = moduleNameString $ moduleName $ ifaceMod iface
+          annots  = show_AnnMap modl $ getAnnMap (fsLit src) $ ifaceTcSource iface
+
+newtype AnnMap = Ann (M.Map Loc (String, String))
+type Loc = SrcSpan
+
+show_AnnMap :: String -> AnnMap -> String
+show_AnnMap modl annMap  = "\n\n" ++ (concatMap ppAnn $ realAnns annMap)
+    where realAnns (Ann m) = dropBadSpans (M.toList m)
+
+          -- Discard UnhelpfulSpans
+          dropBadSpans [] = []
+          dropBadSpans ((RealSrcSpan ss, v):xs) = (ss, v) : dropBadSpans xs
+          dropBadSpans ((UnhelpfulSpan _msg, _v):xs) = dropBadSpans xs
+
+          ppAnn :: (RealSrcSpan, ([Char], [Char])) -> [Char]
+          ppAnn (l, (x,s)) =  x ++ "\n"
+                           ++ modl ++ "\n"
+                           ++ show (srcSpanStartLine l) ++ "\n"
+                           ++ show (srcSpanStartCol l)  ++ "\n"
+                           ++ show (length $ lines s)   ++ "\n"
+                           ++ s ++ "\n\n\n"
+
+---------------------------------------------------------------------------
+-- Extract Annotations ----------------------------------------------------
+---------------------------------------------------------------------------
+
+getAnnMap ::  Data a => FastString -> a -> AnnMap
+getAnnMap src tcm  = Ann $ M.fromList $ canonize $ anns
+  where anns   = [(l, (s, renderId x)) | (l, (s, x)) <- rs ++ ws ]
+        rs     = [(l, (s, x)) | (_,l,s) <- getLocEs tcm, x <- typs s]
+        ws     = [(l, (s, x)) | (s, (x, Just l)) <- ns]
+        ns     = getNames src tcm
+        tm     = M.fromList ns
+        typs s = case s `M.lookup` tm of
+                   Nothing    -> []
+                   Just (x,_) -> [x]
+
+canonize :: (Ord b, Eq a) => [(b, (t, [a]))] -> [(b, (t, [a]))]
+canonize anns = maximumBy cmp $ groupWith fst anns
+  where maximumBy o = catMaybes . fmap (safeHead . L.sortBy o)
+        safeHead [] = Nothing
+        safeHead (x:_) = Just x
+        cmp (_,(_,x1)) (_,(_,x2))
+          | x1 == x2              = EQ
+          | length x1 < length x2 = GT
+          | otherwise             = LT
+
+getLocEs ::  (Data a) => a -> [(HsExpr Id, Loc, String)]
+getLocEs z = [(e, l, stripParens $ unsafePpr e) | L l e <- findLEs z]
+  where stripParens ('(':s)  = stripParens s
+        stripParens s        = stripRParens (reverse s)
+        stripRParens (')':s) = stripRParens s
+        stripRParens s       = reverse s
+
+getNames ::  (Data a) => FastString -> a -> [(String, (Id, Maybe Loc))]
+getNames src z = [(unsafePpr x, (x, idLoc src x)) | x <- findIds z]
+
+renderId :: Id -> String
+renderId = showSDocForUser unsafeGlobalDynFlags neverQualify . pprTyThing showSub . AnId
```

Comment:
I'm using `unsafeGlobalDynFlags` throughout. Since we run after typechecking this seems safe?

Reply:
Please thread `DynFlags` through. Relying on global state breaks my reasoning.